ols/src/query_helpers/__init__.py

```diff
@@ -1,16 +1,52 @@
 """Question validators, statament classifiers, and response generators."""
 
+import logging
+from functools import lru_cache
 from typing import Optional
 
+from langchain.llms.base import LLM
+
+from ols.src.llms.llm_loader import LLMLoader
 from ols.utils import config
 
+logger = logging.getLogger(__name__)
+
 
 class QueryHelper:
     """Base class for query helpers."""
 
     def __init__(
-        self, provider: Optional[str] = None, model: Optional[str] = None
+        self,
+        provider: Optional[str] = None,
+        model: Optional[str] = None,
+        llm_loader: Optional[LLMLoader] = None,
     ) -> None:
         """Initialize query helper."""
+        # NOTE: As signature of this method is evaluated before the config,
+        # is loaded, we cannot use the config directly as defaults and we
+        # need to use those values in the init evaluation.
         self.provider = provider or config.ols_config.default_provider
         self.model = model or config.ols_config.default_model
+        self.llm_loader = llm_loader or LLMLoader
+
+    @property
+    def llm_params(self) -> dict:
+        """Return the parameters for the LLM."""
+        logger.info(f"no LLM parameters are configured for {__class__.__name__}")
+        return {}
+
+    @lru_cache(maxsize=None)
+    def _get_llm(self, provider: str, model: str, **llm_params) -> LLM:
+        loaded_llm = self.llm_loader(provider, model, params=llm_params)
+        return loaded_llm.llm
+
+    def get_llm(self) -> LLM:
```

Comment:
by doing this this way you've made it difficult or impossible for a single query helper to make use of more than one LLM, or use different parameters for the llm.

i don't know if we'll have that use case, but in general i'm not a fan of the way this has been restructured.  It blends too much responsibility between the llm loader and the query helper.

What is wrong with a single utility function on llm_loader that takes the model/provider/model_params as args returns the llm instance to the caller?  The return value can still be cached.

Reply:
Parameters are passed from the respective query helper.
Only single llm per query helper - true.

With the utility function, you mean something like this
```python
@lru_cache(maxsize=None)  
def get_llm(provide, model, params) -> LLM:
    llm_loader = LLMLoader(provider=provide, model=model, params=params)
    return llm_loader.llm
```
in `llm_loader.py`? And you would call it directly in query helper? That is against the dependency injection.

I think it can be fixed like this:

- I'll remove the `get_llm` and `_get_llm` methods from query helper.
  - looking at that now again, you are right, it "tights hands" too much
- Revert to the original way of initializing llm loader, eg. `bare_llm = self.llm_loader(provider, model, params=llm_params)`
- Have caching on the side of `LLMLoader`.

So with that, you will be able to create multiple instances of llm in the query helpers.
WDYT?